polymod/fs/NodeFileSystem.hx

```diff
@@ -0,0 +1,155 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * 
+ */
+ 
+package polymod.fs;
+import haxe.io.Bytes;
+import haxe.io.UInt8Array;
+import js.Browser;
+import js.html.ScriptElement;
+import js.Lib;
+
+class NodeFileSystem
+{
+	// hack to make sure NodeUtils.injectJSCode is called
+	private static var _jsCodeInjected:Bool = injectJSCode();
+	
+	// -----------------------------------------------------------------------------------------------
+	// -----------------------------------------------------------------------------------------------
+	/**
+	 * Injects JS code needed to interact with Node's file system into the head element of the HTML document.
+	 * @return
+	 */
+	private static function injectJSCode():Bool
+	{
+		// array for adding JS text
+		var jsCode:Array<String> = [];
+		
+		// get the node file system
+		jsCode.push("let _nodefs = require('fs')");
+		
+		// utility function for getting directory contents
+		jsCode.push("function getDirectoryContents(path, recursive, dirContents=null)");
+		jsCode.push("{");
+		jsCode.push("	if ( dirContents == null ) {");
+		jsCode.push("		dirContents = [];");
+		jsCode.push("	}");
+		jsCode.push("	if ( isDirectory(path) ) {");
+		jsCode.push("		if ( path.charAt(path.length - 1) != '/' ) {");
+		jsCode.push("			path += '/';");
+		jsCode.push("		}");
+		jsCode.push("		var entries = _nodefs.readdirSync(path, { withFileTypes:true } );");
+		jsCode.push("		for ( var i = 0; i < entries.length; ++i ) {");
+		jsCode.push("			var entryPath = path + entries[i].name;");
+		jsCode.push("			dirContents.push( entryPath );");
+		jsCode.push("			if ( entries[i].isDirectory() && recursive ) {");
+		jsCode.push("				getDirectoryContents( entryPath, true, dirContents );");
+		jsCode.push("			}");
+		jsCode.push("		}");
+		jsCode.push("	}");
+		jsCode.push("	return dirContents;");
+		jsCode.push("}");
+		
+		// functions needed by Polymod
+		jsCode.push("function exists(path) { return _nodefs.existsSync(path); }");
+		jsCode.push("function getStats(path) { return exists(path) ? _nodefs.statSync(path) : null; }");
+		jsCode.push("function isDirectory(path) { var stats = getStats(path); return stats != null && stats.isDirectory(); }");
+		jsCode.push("function getFileContent(path) { return exists(path) ? _nodefs.readFileSync(path, {encoding:'utf8', flag:'r'}) : ''; }");
+		jsCode.push("function getFileBytes(path) { return exists(path) ? Uint8Array.from( _nodefs.readFileSync(path) ) : null; }");
+		jsCode.push("function readDirectory(path) { return getDirectoryContents(path, false, []) }");
+		jsCode.push("function readDirectoryRecursive(path) { return getDirectoryContents(path, true, []) }");
+		
+		// create the script element
+		var scriptElement:ScriptElement = Browser.document.createScriptElement();
+		scriptElement.type = "text/javascript";
+		scriptElement.text = jsCode.join("\n");
+		
```

Comment:
all this jsCode seems to just get turned into one big blob of text, yeah? There's probably a cleaner way to do this -- was there any particular reason for setting it up this way I'm not appreciating?

Reply:
(as in, if it's just a big blob of text I'm not sure we need all the intermediate arrays, I think haxe code probably has a way to let you have a big blob of inline text without having to internally escape all your quotation marks or whatever)

Reply:
I'm open to suggestions on how to format it better.

Since this is an operation that only happens once, I just used an array and a join call to avoid having manually adding newlines in the text so it was relatively easy to read.